File: bindings/java/samples/Env3.java
/**
 * Copyright (C) 2005-2008 Christoph Rupp (chris@crupp.de).
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or 
 * (at your option) any later version.
 *
 * See file COPYING.GPL2 and COPYING.GPL3 for License information.
 *
 */

/*
 * This sample is the implementation of /samples/env3.cpp in Java.
 *
 * It creates an Environment with three Databases - one for the Customers,
 * one for Orders, and a third for managing the 1:n relationship between
 * the other two.
 */

import de.crupp.hamsterdb.*;

public class Env3 {

    /*
     * a Customer class
     */
    private class Env3Customer {
        public Env3Customer(int id, String name) {
            this.id=id;
            this.name=name;
        }

        public Env3Customer(byte[] bid, byte[] bname) {
            this.id=Env3.btoi(bid);
            this.name=new String(bname);
        }
    
        public int id;              /*< the unique Customer ID */
        public String name;         /*< the Customer's name */
        // ...

        public byte[] getKey() {
            // return the id as a byte array
            return Env3.itob(id);
        }

        public byte[] getRecord() {
            // return the id and name as a byte array
            return name.getBytes();
        }
    }

    /*
     * an Order class; it stores the ID of the Customer,
     * and the name of the employee who is assigned to this order
     */
    private class Env3Order {
        public Env3Order(int id, int customer_id, String assignee) {
            this.id=id;
            this.customer_id=customer_id;
            this.assignee=assignee;
        }

        public int id;              /*< the unique Order ID */
        public int customer_id;     /*< the Customer of this Order */
        public String assignee;     /*< the Employee who is assigned to this 
                                      order */
        // ...

        public byte[] getKey() {
            // return the ID as a byte array
            return Env3.itob(id);
        }

        public byte[] getCustomerKey() {
            // return the customer ID as a byte array
            return Env3.itob(customer_id);
        }

        public byte[] getRecord() {
            // return the assignee as a byte array
            return assignee.getBytes();
        }
    }

    /*
     * store an integer in a byte array
     */
    public static byte[] itob(int i) {
        return new byte[] {
            (byte)(i >>> 24),
            (byte)(i >>> 16),
            (byte)(i >>> 8),
            (byte) i};
    }

    /*
     * read an integer from a byte array
     */
    public static int btoi(byte[] b) {
        return (b[0] << 24)
            + ((b[1] & 0xFF) << 16)
            + ((b[2] & 0xFF) << 8)
            + ( b[3] & 0xFF); 
    }

    static final int MAX_DBS           = 3;

    static final short DBNAME_CUSTOMER = 1;
    static final short DBNAME_ORDER    = 2;
    static final short DBNAME_C2O      = 3;  /* C2O: Customer To Order */

    static final int DBIDX_CUSTOMER    = 0;
    static final int DBIDX_ORDER       = 1;
    static final int DBIDX_C2O         = 2;

    private Environment m_env;
    private Database m_db[];
    private Cursor m_cursor[];

    public void init() 
            throws de.crupp.hamsterdb.Error {
        m_env=new Environment();
        m_db=new Database[MAX_DBS];
        m_cursor=new Cursor[MAX_DBS];

        /*
         * create a new Environment file
         */
        m_env.create("test.db");

        /*
         * then create the two Databases in this Environment; each Database
         * has a name - the first is our "customer" Database, the second 
         * is for the "orders"; the third manages our 1:n relation and
         * therefore needs to enable duplicate keys
         */
        m_db[DBIDX_CUSTOMER]=m_env.createDatabase(DBNAME_CUSTOMER);
        m_db[DBIDX_ORDER]   =m_env.createDatabase(DBNAME_ORDER);
        m_db[DBIDX_C2O]     =m_env.createDatabase(DBNAME_C2O, 
                Const.HAM_ENABLE_DUPLICATES);

        /* 
         * create a Cursor for each Database
         */
        for (int i=0; i<MAX_DBS; i++)
            m_cursor[i]=new Cursor(m_db[i]);
    }

    public void close() 
            throws de.crupp.hamsterdb.Error {
        /*
         * close all Cursors, Databases and the Environment
         */
        for (int i=0; i<MAX_DBS; i++)
            m_cursor[i].close();
        for (int i=0; i<MAX_DBS; i++)
            m_db[i].close();
        m_env.close();
    }

    public void run() 
            throws de.crupp.hamsterdb.Error {

        Env3Customer customers[]={
            new Env3Customer(1, "Alan Antonov Corp."),
            new Env3Customer(2, "Barry Broke Inc."),
            new Env3Customer(3, "Carl Caesar Lat."),
            new Env3Customer(4, "Doris Dove Brd.")
        };

        Env3Order orders[]={
            new Env3Order(1, 1, "Joe"),
            new Env3Order(2, 1, "Tom"),
            new Env3Order(3, 3, "Joe"),
            new Env3Order(4, 4, "Tom"),
            new Env3Order(5, 3, "Ben"),
            new Env3Order(6, 3, "Ben"),
            new Env3Order(7, 4, "Chris"),
            new Env3Order(8, 1, "Ben")
        };

        /*
         * insert the customers in the customer table
         *
         * INSERT INTO customers VALUES (1, "Alan Antonov Corp.");
         * INSERT INTO customers VALUES (2, "Barry Broke Inc.");
         * etc
         */
        for (int i=0; i<customers.length; i++) {
            byte[] key=customers[i].getKey();
            byte[] rec=customers[i].getRecord();

            m_db[DBIDX_CUSTOMER].insert(key, rec);
        }

        /*
         * and now the orders in the second Database; contrary to env1, 
         * we only store the assignee, not the whole structure
         *
         * INSERT INTO orders VALUES (1, "Joe");
         * INSERT INTO orders VALUES (2, "Tom");
         */
        for (int i=0; i<orders.length; i++) {
            byte[] key=orders[i].getKey();
            byte[] rec=orders[i].getRecord();

            m_db[DBIDX_ORDER].insert(key, rec);
        }

        /*
         * and now the 1:n relationships; the flag HAM_DUPLICATE creates
         * a duplicate key, if the key already exists
         *
         * INSERT INTO c2o VALUES (1, 1);
         * INSERT INTO c2o VALUES (2, 1);
         * etc
         */
        for (int i=0; i<orders.length; i++) {
            byte[] key=orders[i].getCustomerKey();
            byte[] rec=orders[i].getKey();

            m_db[DBIDX_C2O].insert(key, rec, Const.HAM_DUPLICATE);
        }

        /*
         * now start the query - we want to dump each customer with his
         * orders
         *
         * loop over the customer; for each customer, loop over the 1:n table
         * and pick those orders with the customer id. then load the order
         * and print it
         *
         * the outer loop is similar to 
         * SELECT * FROM customers WHERE 1;
         */
        while (1==1) {
            Env3Customer customer;

            try {
                m_cursor[DBIDX_CUSTOMER].moveNext();
            }
            catch (de.crupp.hamsterdb.Error e) {
                // reached end of Database? 
                if (e.getErrno()==Const.HAM_KEY_NOT_FOUND)
                    break;
                System.out.println("cursor.moveNext failed: "+e);
                return;
            }

            // load the Customer
            customer=new Env3Customer(m_cursor[DBIDX_CUSTOMER].getKey(),
                    m_cursor[DBIDX_CUSTOMER].getRecord());

            // print information about this Customer
            System.out.println("customer "+customer.id+" ('"+
                    customer.name+"')");

            /*
             * loop over the 1:n table
             *
             * before we start the loop, we move the cursor to the
             * first duplicate key
             *
             * SELECT * FROM customers, orders, c2o 
             *   WHERE c2o.customer_id=customers.id AND
             *      c2o.order_id=orders.id;
             */
            try {
                m_cursor[DBIDX_C2O].find(customer.getKey());
            }
            catch (de.crupp.hamsterdb.Error e) {
                // no order for this Customer?
                if (e.getErrno()==Const.HAM_KEY_NOT_FOUND)
                    continue;
                System.out.println("cursor.find failed: "+e);
                return;
            }

            do {
                /* 
                 * load the order; order_id is a byteArray with the ID of the
                 * Order; the record of the item is a byteArray with the
                 * name of the assigned employee
                 *
                 * SELECT * FROM orders WHERE id = order_id;
                 */
                byte[] order_id=m_cursor[DBIDX_C2O].getRecord();
                m_cursor[DBIDX_ORDER].find(order_id);
                String assignee=new String(m_cursor[DBIDX_ORDER].getRecord());

                System.out.println("  order: "+btoi(order_id)+" (assigned to "+
                        assignee+")");
                
                /*
                 * move to the next order which belongs to this customer
                 *
                 * the flag HAM_ONLY_DUPLICATES restricts the cursor 
                 * movement to the duplicate list.
                 */
                try {
                    m_cursor[DBIDX_C2O].move(Const.HAM_CURSOR_NEXT
                            |Const.HAM_ONLY_DUPLICATES);
                }
                catch (de.crupp.hamsterdb.Error e) {
                    // no more order for this customer?
                    if (e.getErrno()==Const.HAM_KEY_NOT_FOUND)
                        break;
                    System.out.println("cursor.moveNext failed: "+e);
                    return;
                }
            } while(1==1);
        }

        /* TODO ... */
        System.out.println("Success!");
    }

	public static void main(String args[]) {
		try {
            Env3 env3=new Env3();
            env3.init();
            env3.run();
            env3.close();
        }
		catch (de.crupp.hamsterdb.Error err) {
			System.out.println("Exception "+err);
		}
	}
}

File: samples/env3.cpp
/*
 * Copyright (C) 2005-2016 Christoph Rupp (chris@crupp.de).
 * All Rights Reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * See the file COPYING for License information.
 */

/**
 * This sample does the same as env2, but uses the C++ API.
 */

#include <iostream>
#include <stdlib.h> /* for exit() */
#include <ups/upscaledb.hpp>

#define MAX_DBS           3

#define DBNAME_CUSTOMER   1
#define DBNAME_ORDER      2
#define DBNAME_C2O        3   /* C2O: Customer To Order */

#define DBIDX_CUSTOMER    0
#define DBIDX_ORDER       1
#define DBIDX_C2O         2

#define MAX_CUSTOMERS     4
#define MAX_ORDERS        8

/* A structure for the "customer" database */
typedef struct {
  uint32_t id;         /* customer id; will be the key of the customer table */
  char name[32];        /* customer name */
  /* ... additional information could follow here */
} customer_t;

/* A structure for the "orders" database */
typedef struct {
  uint32_t id;         /* order id; will be the key of the order table */
  uint32_t customer_id;/* customer id */
  char assignee[32];    /* assigned to whom? */
  /* ... additional information could follow here */
} order_t;

int
run_demo() {
  int i;
  upscaledb::env env;       /* upscaledb environment */
  upscaledb::db db[MAX_DBS];  /* upscaledb database objects */
  upscaledb::cursor cursor[MAX_DBS]; /* a cursor for each database */
  upscaledb::key key, cust_key, ord_key, c2o_key;
  upscaledb::record record, cust_record, ord_record, c2o_record;

  customer_t customers[MAX_CUSTOMERS] = {
    { 1, "Alan Antonov Corp." },
    { 2, "Barry Broke Inc." },
    { 3, "Carl Caesar Lat." },
    { 4, "Doris Dove Brd." }
  };

  order_t orders[MAX_ORDERS] = {
    { 1, 1, "Joe" },
    { 2, 1, "Tom" },
    { 3, 3, "Joe" },
    { 4, 4, "Tom" },
    { 5, 3, "Ben" },
    { 6, 3, "Ben" },
    { 7, 4, "Chris" },
    { 8, 1, "Ben" }
  };

  /* Create a new upscaledb environment */
  env.create("test.db");

  /*
   * Then create the two Databases in this Environment; each Database
   * has a name - the first is our "customer" Database, the second
   * is for the "orders"; the third manages our 1:n relation and
   * therefore needs to enable duplicate keys
   *
   * All database keys are uint32 types.
   */
  ups_parameter_t params[] = {
    {UPS_PARAM_KEY_TYPE, UPS_TYPE_UINT32},
    {0, }
  };

  /*
   * The "mapping" between customers and orders stores uint32 customer IDs
   * as a key and uint32 order IDs as a record
   */
  ups_parameter_t c2o_params[] = {
    {UPS_PARAM_KEY_TYPE, UPS_TYPE_UINT32},
    {UPS_PARAM_RECORD_SIZE, sizeof(uint32_t)},
    {0, }
  };

  db[DBIDX_CUSTOMER] = env.create_db(DBNAME_CUSTOMER, 0, &params[0]);
  db[DBIDX_ORDER]  = env.create_db(DBNAME_ORDER, 0, &params[0]);
  db[DBIDX_C2O]    = env.create_db(DBNAME_C2O,
                  UPS_ENABLE_DUPLICATE_KEYS, &c2o_params[0]);

  /* Create a cursor for each database */
  for (i = 0; i < MAX_DBS; i++)
    cursor[i].create(&db[i]);

  /*
   * Insert the customers in the customer table
   *
   * INSERT INTO customers VALUES (1, "Alan Antonov Corp.");
   * INSERT INTO customers VALUES (2, "Barry Broke Inc.");
   * etc
   */
  for (i = 0; i < MAX_CUSTOMERS; i++) {
    key.set_size(sizeof(int));
    key.set_data(&customers[i].id);

    record.set_size(sizeof(customer_t));
    record.set_data(&customers[i]);

    db[0].insert(&key, &record);
  }

  /*
   * And now the orders in the second database; contrary to env1,
   * we only store the assignee, not the whole structure
   *
   * INSERT INTO orders VALUES (1, "Joe");
   * INSERT INTO orders VALUES (2, "Tom");
   */
  for (i = 0; i < MAX_ORDERS; i++) {
    key.set_size(sizeof(int));
    key.set_data(&orders[i].id);

    record.set_size(sizeof(orders[i].assignee));
    record.set_data(orders[i].assignee);

    db[1].insert(&key, &record);
  }

  /*
   * And now the 1:n relationships; the flag UPS_DUPLICATE creates
   * a duplicate key, if the key already exists
   *
   * INSERT INTO c2o VALUES (1, 1);
   * INSERT INTO c2o VALUES (2, 1);
   * etc
   */
  for (i = 0; i < MAX_ORDERS; i++) {
    key.set_size(sizeof(int));
    key.set_data(&orders[i].customer_id);

    record.set_size(sizeof(int));
    record.set_data(&orders[i].id);

    db[2].insert(&key, &record, UPS_DUPLICATE);
  }

  /*
   * Now start the query - we want to dump each customer with his
   * orders
   *
   * loop over the customer; for each customer, loop over the 1:n table
   * and pick those orders with the customer id. then load the order
   * and print it
   *
   * the outer loop is similar to
   * SELECT * FROM customers WHERE 1;
   */
  while (1) {
    customer_t *customer;

    try {
      cursor[0].move_next(&cust_key, &cust_record);
    }
    catch (upscaledb::error &e) {
      /* reached end of the database? */
      if (e.get_errno() == UPS_KEY_NOT_FOUND)
        break;
      else {
        std::cerr << "cursor.move_next() failed: " << e.get_string()
              << std::endl;
        return (-1);
      }
    }

    customer = (customer_t *)cust_record.get_data();

    /* print the customer id and name */
    std::cout << "customer " << customer->id << " ('"
          << customer->name << "')" << std::endl;

    /*
     * Loop over the 1:n table
     *
     * before we start the loop, we move the cursor to the
     * first duplicate key
     *
     * SELECT * FROM customers, orders, c2o
     *   WHERE c2o.customer_id=customers.id AND
     *    c2o.order_id=orders.id;
     */
    c2o_key.set_data(&customer->id);
    c2o_key.set_size(sizeof(int));

    try {
      cursor[2].find(&c2o_key);
    }
    catch (upscaledb::error &e) {
      if (e.get_errno() == UPS_KEY_NOT_FOUND)
        continue;
      else {
        std::cerr << "cursor.find() failed: " << e.get_string()
              << std::endl;
        return (-1);
      }
    }

    /* get the record of this database entry */
    cursor[2].move(0, &c2o_record);

    do {
      int order_id;

      order_id = *(int *)c2o_record.get_data();
      ord_key.set_data(&order_id);
      ord_key.set_size(sizeof(int));

      /*
       * Load the order
       * SELECT * FROM orders WHERE id = order_id;
       */
      ord_record = db[1].find(&ord_key);

      std::cout << "  order: " << order_id << " (assigned to "
            << (char *)ord_record.get_data() << ")" << std::endl;

      /*
       * the flag UPS_ONLY_DUPLICATES restricts the cursor
       * movement to the duplicate list.
       */
      try {
        cursor[2].move(&c2o_key, &c2o_record,
              UPS_CURSOR_NEXT | UPS_ONLY_DUPLICATES);
      }
      catch (upscaledb::error &e) {
        /* reached end of the database? */
        if (e.get_errno() == UPS_KEY_NOT_FOUND)
          break;
        else {
          std::cerr << "cursor.move() failed: " << e.get_string()
              << std::endl;
          return (-1);
        }
      }

    } while (1);
  }

  /*
   * we're done! no need to cleanup, the destructors will prevent memory
   * leaks
   */
  std::cout << "success!" << std::endl;
  return (0);
}

int
main(int argc, char **argv)
{
  try {
    return (run_demo());
  }
  catch (upscaledb::error &e) {
    std::cerr << "run_demo() failed with unexpected error "
          << e.get_errno() << " ('"
          << e.get_string() << "')" << std::endl;
    return (-1);
  }
}
